Fully implement flash.geom.ColorTransform

CUB3D opened the pull request:
Provides a full implementation of #302

Herschel commented on core/src/avm1/globals/color_transform.rs:
```diff
@@ -0,0 +1,299 @@
+    avm: &mut Avm1<'gc>,
+    context: &mut UpdateContext<'_, 'gc, '_>,
+) -> Result<(), Error<'gc>> {
+    object.set("redMultiplier", color_transform.r_mult.into(), avm, context)?;
+    object.set(
+        "greenMultiplier",
+        color_transform.g_mult.into(),
+        avm,
+        context,
+    )?;
+    object.set(
+        "blueMultiplier",
+        color_transform.b_mult.into(),
+        avm,
+        context,
+    )?;
+    object.set(
+        "alphaMultiplier",
+        color_transform.a_mult.into(),
+        avm,
+        context,
+    )?;
+    object.set("redOffset", color_transform.r_add.into(), avm, context)?;
+    object.set("greenOffset", color_transform.g_add.into(), avm, context)?;
+    object.set("blueOffset", color_transform.b_add.into(), avm, context)?;
+    object.set("alphaOffset", color_transform.a_add.into(), avm, context)?;
```
Unfortunately we shouldn't round-trip between the engine `ColorTransform` object and the object properties, because this will lose precision from the f64 to f32 (and this will get worse once #193 is fixed). The values should just be stored as properties on its own object native object type, and then we can convert to `color_transform::ColorTransform` only when it is needed.

Try this:
```
var ct = new flash.geom.ColorTransform();
ct.redMultiplier = 8589934592;
trace(ct);
```
Expected: `redMultiplier=8589934592` This PR: `redMultiplier=8589935000`.

This does mean we have to duplicate all of the `concat` logic between engine and this code, but that unfortunately seems to be necessary. (Matrix has the same issue).

Herschel commented on core/src/avm1/globals/color_transform.rs:
```diff
@@ -0,0 +1,299 @@
+    this: Object<'gc>,
+    args: &[Value<'gc>],
+) -> Result<ReturnValue<'gc>, Error<'gc>> {
+    let new_rgb = args
+        .get(0)
+        .unwrap_or(&Value::Undefined)
+        .coerce_to_f64(avm, context)? as u64;
```
This should be `coerce_to_u32` because it takes care of the necessary wrapping behavior. If `f64` is huge, Rust (recently) defines float-to-int casts to saturate, which won't produce the correct value. Try:

```
var ct = new flash.geom.ColorTransform();
var n = 36893488147419103232;
ct.rgb = n;
trace(ct);
```

Flash:
`(redMultiplier=0, greenMultiplier=0, blueMultiplier=0, alphaMultiplier=1, redOffset=0, greenOffset=0, blueOffset=0, alphaOffset=0)`
Ruffle:
`(redMultiplier=0, greenMultiplier=0, blueMultiplier=0, alphaMultiplier=1, redOffset=255, greenOffset=255, blueOffset=255, alphaOffset=0)
`

Herschel commented on core/src/avm1/globals/color_transform.rs:
```diff
@@ -0,0 +1,299 @@
+        Some(fn_proto),
+    );
+
+    object.add_property(
+        gc_context,
+        "rgb",
+        Executable::Native(get_rgb),
+        Some(Executable::Native(set_rgb)),
+        EnumSet::empty(),
+    );
+
```
It seems like the `redMultiplier` etc. need to be virtual properties as well:
`trace(flash.geom.ColorTransform.prototype.hasOwnProperty("redMultiplier"));`
And these setters perform some validation on the values (turn weirdos into "NaN"),

Herschel commented on core/src/avm1/globals/color_transform.rs:
```diff
@@ -0,0 +1,299 @@
+    let ct = object_to_color_transform(this, avm, context)?;
+
+    let formatted = format!("(redMultiplier={}, greenMultiplier={}, blueMultiplier={}, alphaMultiplier={}, redOffset={}, greenOffset={}, blueOffset={}, alphaOffset={})",
+            ct.r_mult,
+            ct.g_mult,
+            ct.b_mult,
+            ct.a_mult,
+            ct.r_add,
+            ct.g_add,
+            ct.b_add,
+            ct.a_add
```
This should probably call `coerce_to_string` on all of these to get the proper numeric formatting.

Herschel commented:
Also, it'd be nice to add some tests for this in `swf-tests` (if you don't have access to the Flash IDE, I can take care of this!).

Dinnerbone commented:
Please let me know if you need help rebasing this with the recent changes on master!

CUB3D commented:
Pretty sure that's everything addressed now, I added the tests you suggested to the ones I was using previously and added them to the regression set. If anything else needs changing, just let me know.
